### ISSUE TITLE
shader: Optimize Kage program to use more conditional operators (`?:`)

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#1836]
IIUC, conditional operators are more efficient than `if` branches. It might be hard to do this optimization though...

**Comment by Zyko0** [comment]
I read that in some shader languages it can be an optimization, but in some others it's actually just a syntax which still does regular branching
Is there any documentation that confirms this, I'm quite curious

**Comment by hajimehoshi** [comment]
https://forum.unity.com/threads/correct-use-of-unity_branch.476804/

`UNITY_BRANCH` and `UNITY_FLATTEN` can differ for example.

**Comment by Zyko0** [comment]
Yes but you mentioned `?:` ternary operator specifically, and I'm wondering if in glsl, this still translates into `if/else` or not
Like if the boolean is dynamic, different shader invocations might still take different execution "paths" whether it's `if/else` of `?:`, right ?

I didn't know about that unity thing, does that mean you are thinking about providing `?:` as a `flatten` operator ?

**Comment by hajimehoshi** [comment]
> Yes but you mentioned ?: ternary operator specifically, and I'm wondering if in glsl, this still translates into if/else or not
Like if the boolean is dynamic, different shader invocations might still take different execution "paths" whether it's if/else of ?:, right ?

TBH I don't have a good insight in this, but I think there can be a hardware where (B) works faster than (A), as a compiler might be able to easily optimize this as each branch in a conditional operator cannot have any side effect:

```
// (A)
float a;
if (foo) {
  a = 1.0;
} else {
  a = 2.0;
}
```

```
// (B)
a = foo ? 1.0 : 2.0;
```

Let me investigate more.

> I didn't know about that unity thing, does that mean you are thinking about providing ?: as a flatten operator ?

No it doesn't. This should be done in Kage compiler automatically!

**Comment by superloach** [comment]
something as simple as
```
// (A)
float a;
if (foo) {
  a = 1.0;
} else {
  a = 2.0;
}
```
may actually be optimised into something like
```
// (C)
float a = 1.0 + float(foo);
```
or whatever the glsl equivalent would be, assuming truthy values are 1
